File: mcp/sketchup-managed-mcp/runtime-support/professional-sketchup-modeling/scripts/read_instance_layout.rb
require 'json'

# Read-only occurrence traversal. No open/save/selection/camera/attribute/model writes.
module PipClawInstanceReadback
  extend self
  def container?(e)
    e.is_a?(Sketchup::Group) || e.is_a?(Sketchup::ComponentInstance)
  end
  def identity(e)
    e.persistent_id.to_s
  end
  def matrix(t)
    a=t.to_a
    # SketchUp column-major inches -> row-major matrix with metre translations.
    (0..3).map { |r| (0..3).map { |c| (c==3 && r<3) ? a[c*4+r].to_f*0.0254 : a[c*4+r].to_f } }
  end
  def run(options)
    model=Sketchup.active_model
    expected=options.fetch('expected_path')
    raise 'Expected saved document path required' if expected.to_s.empty? || model.path.to_s.empty?
    same=begin File.identical?(expected,model.path); rescue; File.expand_path(expected).casecmp(File.expand_path(model.path)).zero?; end
    raise 'Active document does not match expected_path; no document was opened' unless same
    limit=options.fetch('max_instances',1000)
    depth_limit=options.fetch('max_depth',12)
    raise 'Invalid traversal limits' unless limit.is_a?(Integer) && limit.between?(1,5000) && depth_limit.is_a?(Integer) && depth_limit.between?(1,32)
    before={'path'=>model.path.to_s,'guid'=>model.guid.to_s,'modified'=>model.modified?}
    scope=options.fetch('scope_path',[])
    raise 'scope_path must be an array of persistent ID strings' unless scope.is_a?(Array) && scope.length<=32 && scope.all? { |x| x.is_a?(String) && x.match(/\A[0-9]+\z/) }
    entities=model.entities;chain=[];world=Geom::Transformation.new;anc=[];prefix=[]
    scope.each do |pid|
      e=entities.find { |x| container?(x) && identity(x)==pid }
      raise 'scope_path not found in active occurrence hierarchy' unless e
      chain << matrix(e.transformation);world=world*e.transformation
      prefix << identity(e);anc << e.definition.object_id;entities=e.definition.entities
    end
    rows=[];truncated=false;reasons=[];stack=[]
    entities.each do |e|
      next unless container?(e)
      if stack.length>=limit+1;truncated=true;reasons << 'pending_limit';break;end
      stack << [e,prefix,world,chain,anc,1]
    end
    until stack.empty?
      if rows.length>=limit;truncated=true;reasons << 'max_instances';break;end
      e,parent,t,parents,ancestors,depth=stack.pop
      local=e.transformation;wt=t*local;ids=parent+[identity(e)];wm=matrix(wt);axes=(0..2).map { |j| (0..2).map { |i| wm[i][j] } }
      a,b,c=axes;det=a[0]*(b[1]*c[2]-b[2]*c[1])-b[0]*(a[1]*c[2]-a[2]*c[1])+c[0]*(a[1]*b[2]-a[2]*b[1])
      rows << {'entity_path'=>ids.join('/'),'definition_id'=>e.definition.guid.to_s,'definition_name'=>e.definition.name.to_s,'instance_name'=>e.name.to_s,
        'kind'=>e.is_a?(Sketchup::Group) ? 'group' : 'component','origin'=>wm[0..2].map { |r| r[3] },'world_matrix'=>wm,'parent_to_local_chain'=>parents+[matrix(local)],
        'axis_lengths'=>axes.map { |v| Math.sqrt(v.inject(0.0) { |sum,x| sum+x*x }) },'determinant'=>det,'hidden'=>e.hidden?, 'tag'=>e.layer.name.to_s,'tag_visible'=>e.layer.visible?}
      children=[]
      e.definition.entities.each do |x|
        next unless container?(x)
        if children.length>=limit+1;truncated=true;reasons << 'pending_limit';break;end
        children << x
      end
      unless children.empty?
        if depth>=depth_limit || ancestors.include?(e.definition.object_id)
          truncated=true;reasons << (depth>=depth_limit ? 'max_depth' : 'definition_cycle');next
        end
        children.each do |child|
          if stack.length>=limit+1;truncated=true;reasons << 'pending_limit';break;end
          stack << [child,ids,wt,parents+[matrix(local)],ancestors+[e.definition.object_id],depth+1]
        end
      end
    end
    after={'path'=>model.path.to_s,'guid'=>model.guid.to_s,'modified'=>model.modified?}
    raise 'Document identity/modified flag changed during readback' unless before==after
    {'schema_version'=>1,'scope'=>'active_document_occurrences','source'=>before,'source_after'=>after,'units'=>'m','coordinate_frame'=>'world:model-origin',
      'scope_path'=>scope,'instances'=>rows,'truncated'=>truncated,'truncation_reasons'=>reasons.uniq,'geometry_readback'=>'instance_transforms_only',
      'visibility_scope'=>'all_containers_including_hidden; inherited/scene visibility not resolved','surface_contact'=>'unverified'}
  end
end

File: skill/professional-sketchup-modeling/scripts/read_instance_layout.py
"""Normalize the existing read-only C API audit to occurrence/world-transform rows.
Snapshot-local child-index paths and definition IDs; never use them as live persistent IDs.
"""
import argparse
import hashlib
import json
from pathlib import Path


def multiply(a,b):return [[sum(a[i][k]*b[k][j] for k in range(4)) for j in range(4)] for i in range(4)]

def local_matrix(values):
    if not isinstance(values,list) or len(values)!=16:raise ValueError('Expected SketchUp flat transform')
    import math
    if any(type(v) not in (int,float) or not math.isfinite(v) for v in values):raise ValueError('Nonfinite transform')
    m=[[values[j*4+i] for j in range(4)] for i in range(4)]
    if m[3]!=[0,0,0,1]:raise ValueError('Expected affine transform')
    for i in range(3):m[i][3]*=.0254
    return m


def extract(data,scope='ROOT',limit=1000,depth_limit=12):
    if type(limit) is not int or not 1<=limit<=5000 or type(depth_limit) is not int or not 1<=depth_limit<=32:raise ValueError('Invalid traversal limit')
    definitions=data['definitions'];node=data['root'];chain=[];world=[[int(i==j) for j in range(4)] for i in range(4)]
    parts=scope.split('/')
    if parts[0]!='ROOT':raise ValueError('scope must start with ROOT')
    ancestors=[]
    for part in parts[1:]:
        if not part.isdecimal():raise ValueError('Expected snapshot child index')
        child=node['children'][int(part)];local=local_matrix(child['transform']);chain.append(local);world=multiply(world,local)
        ancestors.append(child['definition']);node=definitions[child['definition']]
    stack=[(c,f'{scope}/{i}',world,chain,ancestors,1) for i,c in enumerate(node['children'])]
    rows=[];reasons=[]
    while stack:
        if len(rows)>=limit:reasons.append('max_instances');break
        child,path,parent,parents,anc,depth=stack.pop();key=child['definition'];n=definitions[key]
        local=local_matrix(child['transform']);wm=multiply(parent,local);newchain=parents+[local]
        a,b,c=[[wm[i][j] for i in range(3)] for j in range(3)]
        det=a[0]*(b[1]*c[2]-b[2]*c[1])-b[0]*(a[1]*c[2]-a[2]*c[1])+c[0]*(a[1]*b[2]-a[2]*b[1])
        rows.append({'entity_path':path,'definition_id':key,'definition_name':n.get('name'),'instance_name':child.get('name'),'origin':[wm[i][3] for i in range(3)],'world_matrix':wm,'parent_to_local_chain':newchain,'determinant':det,'axis_lengths':[sum(v*v for v in x)**.5 for x in [a,b,c]],'hidden':child.get('hidden'),'tag':child.get('layer')})
        if n['children']:
            if depth>=depth_limit or key in anc:reasons.append('max_depth' if depth>=depth_limit else 'definition_cycle');continue
            stack.extend((ch,f'{path}/{i}',wm,newchain,anc+[key],depth+1) for i,ch in enumerate(n['children']))
    return {'schema_version':1,'scope':'c_api_audit_snapshot_occurrences','source':data['source'],'source_sha256':data['source_sha256'],'units':'m','coordinate_frame':'world:model-origin','scope_path':scope,'instances':rows,'truncated':bool(reasons),'truncation_reasons':sorted(set(reasons)),'geometry_readback':'source_audit_snapshot_not_active_document','identity_scope':'child_index_paths_and_definition_keys_valid_only_for_this_audit','surface_contact':'unverified'}


if __name__=='__main__':
    p=argparse.ArgumentParser(description=__doc__);p.add_argument('--audit',type=Path,required=True);p.add_argument('--output',type=Path,required=True);p.add_argument('--scope',default='ROOT');p.add_argument('--limit',type=int,default=1000);p.add_argument('--depth',type=int,default=12);a=p.parse_args()
    if a.output.exists() or a.output.resolve()==a.audit.resolve():p.error('Use new output file')
    raw=a.audit.read_bytes();data=json.loads(raw.decode('utf-8-sig'));result=extract(data,a.scope,a.limit,a.depth)
    result['audit_sha256']=hashlib.sha256(raw).hexdigest();result['audit_path']=str(a.audit.resolve())
    a.output.write_text(json.dumps(result,ensure_ascii=False,indent=2,allow_nan=False),encoding='utf-8')
